demo/src/r2_12_1_ads/java/com/google/android/exoplayer2/demo/PlayerActivity.java

```diff
@@ -0,0 +1,607 @@
+    }
+    boolean haveStartPosition = startWindow != C.INDEX_UNSET;
+    if (haveStartPosition) {
+      player.seekTo(startWindow, startPosition);
+    }
+    player.setMediaItems(mediaItems, /* resetPosition= */ !haveStartPosition);
+    player.prepare();
+    updateButtonVisibility();
+    return true;
+  }
+
+  private List<MediaItem> createMediaItems(Intent intent) {
+    String action = intent.getAction();
+    boolean actionIsListView = IntentUtil.ACTION_VIEW_LIST.equals(action);
+    if (!actionIsListView && !IntentUtil.ACTION_VIEW.equals(action)) {
+      showToast(getString(R.string.unexpected_intent_action, action));
+      finish();
+      return Collections.emptyList();
+    }
+
+    List<MediaItem> mediaItems =
+        createMediaItems(intent, DemoUtil.getDownloadTracker(/* context= */ this));
+    boolean hasAds = false;
+    for (int i = 0; i < mediaItems.size(); i++) {
+      MediaItem mediaItem = mediaItems.get(i);
+
+      if (!Util.checkCleartextTrafficPermitted(mediaItem)) {
+        showToast(R.string.error_cleartext_not_permitted);
+        return Collections.emptyList();
+      }
+      if (Util.maybeRequestReadExternalStoragePermission(/* activity= */ this, mediaItem)) {
+        // The player will be reinitialized if the permission is granted.
+        return Collections.emptyList();
+      }
+
+      MediaItem.DrmConfiguration drmConfiguration =
+          checkNotNull(mediaItem.playbackProperties).drmConfiguration;
+      if (drmConfiguration != null) {
+        if (Util.SDK_INT < 18) {
+          showToast(R.string.error_drm_unsupported_before_api_18);
+          finish();
+          return Collections.emptyList();
+        } else if (!FrameworkMediaDrm.isCryptoSchemeSupported(drmConfiguration.uuid)) {
+          showToast(R.string.error_drm_unsupported_scheme);
+          finish();
+          return Collections.emptyList();
+        }
+      }
+      hasAds |= mediaItem.playbackProperties.adTagUri != null;
+    }
+    if (!hasAds) {
+      releaseAdsLoader();
+    }
+    return mediaItems;
+  }
+
+  private AdsLoader getAdsLoader(Uri adTagUri) {
+    if (mediaItems.size() > 1) {
+      showToast(R.string.unsupported_ads_in_playlist);
+      releaseAdsLoader();
+      return null;
+    }
+    if (!adTagUri.equals(loadedAdTagUri)) {
+      releaseAdsLoader();
+      loadedAdTagUri = adTagUri;
+    }
+    // The ads loader is reused for multiple playbacks, so that ad playback can resume.
+    if (adsLoader == null) {
+      adsLoader = new ImaAdsLoader.Builder(/* context= */ this)
+              /*
+               * This replace monitor monitorImaAdsLoader method because in this version ImaAdsLoader
+               * will create google.v3.AdsLoader on adRequest, this means that monitorImaAdsLoader
+               * Will always receive null pointer and will be unable to recieve add events.
+               */
+              .setAdErrorListener( muxStats.getAddErrorEventListener() )
```

Comment:
Will need to update this when we rename the methods to `getAdErrorEventListener` for instance